### PR TITLE
Add solutions and tests for challenge_16

### LINKED ISSUE #16: Challenge: Hide a Credit Card Number
# **Challenge: Hide a Credit Card Number**

**Description**

The "Hide a Credit Card Number" challenge involves writing a function to mask all but the last four digits of a credit card number. This is useful for displaying sensitive information in a secure manner. The masked digits should be replaced with the `*` character.

Write a function that takes a credit card number as a string and returns the masked version of the number, with only the last four digits visible.

**For example:**

```
Input: '1234567812345678'
Output: '************5678'
```

**Requirements:**

The input will always be a valid credit card number as a string.
The function should preserve the last four digits.
It should handle inputs of varying lengths (e.g., short card numbers).

**Example**

```
hide_credit_card('1234567812345678')
# Output: '************5678'

hide_credit_card('9876543210987654')
# Output: '************7654'

hide_credit_card('1234')
# Output: '1234' (no masking needed if the number has only 4 or fewer digits)
```

**Testing**

Develop unit tests to validate the function's correctness. Include cases for:

- Regular 16-digit credit card numbers.
- Short numbers (e.g., fewer than 8 digits).
- Edge case: Exactly 4 digits (no masking needed).

**Helpful Links**
- [String Slicing in Python](https://www.geeksforgeeks.org/string-slicing-in-python/)
- [GeeksforGeeks: Python String Methods](https://www.geeksforgeeks.org/python-string-methods/)

### PR DESCRIPTION
**Pull request by hectordacb**
---
name: solution review 
about: A template PR for code review with a checklist
---

## Behavior
# **Challenge: Hide a Credit Card Number**

**Description**

The "Hide a Credit Card Number" challenge involves writing a function to mask all but the last four digits of a credit card number. This is useful for displaying sensitive information in a secure manner. The masked digits should be replaced with the `*` character.

Write a function that takes a credit card number as a string and returns the masked version of the number, with only the last four digits visible.

**For example:**

```
Input: '1234567812345678'
Output: '************5678'
```

**Requirements:**

The input will always be a valid credit card number as a string.
The function should preserve the last four digits.
It should handle inputs of varying lengths (e.g., short card numbers).

**Example**

```
hide_credit_card('1234567812345678')
# Output: '************5678'

hide_credit_card('9876543210987654')
# Output: '************7654'

hide_credit_card('1234')
# Output: '1234' (no masking needed if the number has only 4 or fewer digits)
```

**Testing**

Develop unit tests to validate the function's correctness. Include cases for:

- Regular 16-digit credit card numbers.
- Short numbers (e.g., fewer than 8 digits).
- Edge case: Exactly 4 digits (no masking needed).

**Helpful Links**
- [String Slicing in Python](https://www.geeksforgeeks.org/string-slicing-in-python/)
- [GeeksforGeeks: Python String Methods](https://www.geeksforgeeks.org/python-string-methods/)
### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by likechrisss — CHANGES_REQUESTED** [review]
Only one change is needed to make it perfect. Other than that, impressive work Hector.

**Comment by hectordacb** [comment]
Hi @likechrisss, thanks for the review 😊. Could you please tell me what the requested change is?

File: solutions/tests/challenge_16/test_hide_credit_card_number.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        with self.assertRaises(ValueError):
+            hide_credit_card("")
+
+    def test_non_string_input_raises_value_error(self):
+        """Test with non-string input."""
+        with self.assertRaises(ValueError):
+            hide_credit_card(12345678)
+
+
+if __name__ == "__main__":
```
Comment (likechrisss) [inline solutions/tests/challenge_16/test_hide_credit_card_number.py]: This line is not needed since the test is run using a test discovery tool like: python -m unittest discover

Comment (hectordacb) [reply solutions/tests/challenge_16/test_hide_credit_card_number.py]: It's done now

**Review by likechrisss — APPROVED** [review]
Everything looks good to me.
